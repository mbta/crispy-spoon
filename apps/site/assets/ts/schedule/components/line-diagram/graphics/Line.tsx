import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import {
  isBranchingNode,
  isEndNode,
  isMergeNode,
  isStartNode,
  longestPath
} from "../../../../helpers/stop-tree";
import { hasAnActiveDiversion } from "../../../../models/alert";
import { Alert } from "../../../../__v3api";
import { StopId, StopTree } from "../../__schedule";
import { branchPosition } from "../line-diagram-helpers";
import {
  BASE_LINE_WIDTH,
  BRANCH_LINE_WIDTH,
  BRANCH_SPACING,
  CoordState,
  StopCoord
} from "./graphic-helpers";

interface Props {
  stopTree: StopTree;
  fromId: StopId;
  toId: StopId;
  alerts: Alert[];
}

const isOnPrimaryBranch = (stopTree: StopTree, stopId: StopId): boolean =>
  longestPath(stopTree).includes(stopId);

const isMiddleBranchStop = (stopTree: StopTree, stopId: StopId): boolean =>
  !isOnPrimaryBranch(stopTree, stopId) &&
  !isStartNode(stopTree, stopId) &&
  !isEndNode(stopTree, stopId) &&
  !isMergeNode(stopTree, stopId) &&
  !isBranchingNode(stopTree, stopId);

const lineProps = (
  stopTree: StopTree,
  fromId: StopId,
  toId: StopId,
  fromCoords: StopCoord,
  toCoords: StopCoord,
  alerts: Alert[]
): React.SVGProps<SVGLineElement> => {
  const x =
    BRANCH_SPACING * (branchPosition(stopTree, fromId) - 1) +
    BASE_LINE_WIDTH +
    1;
  const [, y1] = fromCoords;
  const [, y2] = toCoords;

  const strokeWidth =
    isMiddleBranchStop(stopTree, fromId) && isMiddleBranchStop(stopTree, toId)
      ? BRANCH_LINE_WIDTH
      : BASE_LINE_WIDTH;

  const strokeProp =
    hasAnActiveDiversion(fromId, alerts) && hasAnActiveDiversion(toId, alerts)
      ? { stroke: "url(#diagonalHatch)" }
      : {};

  return {
    key: `${fromId}-${toId}`,
    className: "line-diagram-svg__line",
    strokeWidth: `${strokeWidth}px`,
    x1: `${x}px`,
    x2: `${x}px`,
    y1: `${y1}px`,
    y2: `${y2}px`,
    ...strokeProp
  };
};

const Line = ({
  stopTree,
  fromId,
  toId,
  alerts
}: Props): ReactElement | null => {
  const fromCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[fromId]
  );
  const toCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[toId]
  );
  if (!fromCoords || !toCoords) return null;

  if (isMergeNode(stopTree, toId) && !isOnPrimaryBranch(stopTree, fromId))
    return null;

  return (
    <line
      {...lineProps(stopTree, fromId, toId, fromCoords, toCoords, alerts)}
    />
  );
};

const SimpleLine = ({
  fromId,
  toId,
  alerts
}: Omit<Props, "stopTree">): ReactElement | null => {
  const fromCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[fromId]
  );
  const toCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[toId]
  );
  if (!fromCoords || !toCoords) return null;

  const x = BASE_LINE_WIDTH + 1;
  const [, y1] = fromCoords;
  const [, y2] = toCoords;
  const strokeProp =
    hasAnActiveDiversion(fromId, alerts) && hasAnActiveDiversion(toId, alerts)
      ? { stroke: "url(#diagonalHatch)" }
      : {};
  return (
    <line
      className="line-diagram-svg__line"
      strokeWidth={`${BASE_LINE_WIDTH}px`}
      x1={`${x}px`}
      x2={`${x}px`}
      y1={`${y1}px`}
      y2={`${y2}px`}
      {...strokeProp}
    />
  );
};

export { Line, SimpleLine };
