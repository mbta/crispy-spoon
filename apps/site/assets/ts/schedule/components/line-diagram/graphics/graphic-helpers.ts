export const CIRC_RADIUS = 4;
export const CIRC_DIAMETER = CIRC_RADIUS * 2;
export const BASE_LINE_WIDTH = CIRC_DIAMETER + 2;
export const BRANCH_LINE_WIDTH = BASE_LINE_WIDTH + 6;
export const BRANCH_SPACING = 16;
export const MERGE_RADIUS = 6;
