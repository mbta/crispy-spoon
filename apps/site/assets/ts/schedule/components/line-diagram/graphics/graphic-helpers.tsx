import React from "react";
import { AnyAction, Reducer, Store, createStore } from "redux";
import { LineDiagramStop } from "../../__schedule";

export const CIRC_RADIUS = 4;
export const CIRC_DIAMETER = CIRC_RADIUS * 2;
export const BASE_LINE_WIDTH = CIRC_DIAMETER + 2;
export const BRANCH_LINE_WIDTH = BASE_LINE_WIDTH + 6;
export const BRANCH_SPACING = 16;
export const MERGE_RADIUS = 6;

export type StopCoord = [number, number];
export interface CoordState {
  [key: string]: StopCoord | null;
}

export interface CoordAction extends AnyAction {
  stop: string;
  coords: StopCoord | null;
}

export const coordReducer: Reducer<CoordState, CoordAction> = (
  allCoordinates: CoordState | undefined,
  action: CoordAction
): CoordState => {
  const { stop, coords } = action;
  return {
    ...allCoordinates,
    [stop]: coords
  };
};

export const createLineDiagramCoordStore = (
  lineDiagram: LineDiagramStop[]
): Store<CoordState, CoordAction> => {
  const stopIds = lineDiagram.map(ld => ld.route_stop.id);
  const initialCoordState: CoordState = stopIds.reduce(
    (state, id) => ({ ...state, [id]: null }),
    {}
  );
  const store = createStore(
    coordReducer,
    initialCoordState,
    /* eslint-disable no-underscore-dangle, @typescript-eslint/no-explicit-any */
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    /* eslint-enable no-underscore-dangle, @typescript-eslint/no-explicit-any */
  );

  return store;
};

export const DiagonalHatchPattern = (
  id?: string | undefined
): React.SVGProps<SVGPatternElement> => (
  <pattern
    id={id ?? "diagonalHatch"}
    width="10"
    height="10"
    patternTransform="rotate(45 0 0)"
    patternUnits="userSpaceOnUse"
  >
    <line
      className="line-diagram-svg__line"
      y2={CIRC_DIAMETER}
      strokeWidth={`${BRANCH_LINE_WIDTH}px`}
    />
  </pattern>
);
