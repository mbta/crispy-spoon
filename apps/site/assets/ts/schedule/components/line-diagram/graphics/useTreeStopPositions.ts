import { useCallback, useEffect, useRef } from "react";
import { useDispatch } from "react-redux";
import { createStore, Store } from "redux";
import { stopIds } from "../../../../helpers/stop-tree";
import { RouteStop, StopId, StopTree } from "../../__schedule";
import { branchPosition } from "../line-diagram-helpers";
import {
  BASE_LINE_WIDTH,
  BRANCH_SPACING,
  CoordAction,
  coordReducer,
  CoordState
} from "./graphic-helpers";

export type RefMap = Map<string, HTMLElement | null>;

const isRouteStopList = (
  stopOrList: StopTree | RouteStop[]
): stopOrList is RouteStop[] => Array.isArray(stopOrList);

export const createStopTreeCoordStore = (
  stopTreeOrList: StopTree | RouteStop[]
): Store<CoordState, CoordAction> => {
  const ids: StopId[] = isRouteStopList(stopTreeOrList)
    ? stopTreeOrList.map(rs => rs.id)
    : stopIds(stopTreeOrList);

  const initialCoordState: CoordState = ids.reduce(
    (acc, id) => ({ ...acc, [id]: null }),
    {}
  );
  const store = createStore(
    coordReducer,
    initialCoordState,
    /* eslint-disable no-underscore-dangle, @typescript-eslint/no-explicit-any */
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    /* eslint-enable no-underscore-dangle, @typescript-eslint/no-explicit-any */
  );

  return store;
};

const xCoordForStop = (stopTree: StopTree, id: StopId): number =>
  BRANCH_SPACING * (branchPosition(stopTree, id) - 1) + BASE_LINE_WIDTH + 1;

export default function useTreeStopPositions(
  stopTreeOrList: StopTree | RouteStop[]
): [RefMap, () => void] {
  const stopRefsMap = useRef(new Map() as RefMap);
  const dispatchStopCoords = useDispatch();

  const updateAllStops = useCallback((): void => {
    stopRefsMap.current.forEach((el, stopId) => {
      const x = isRouteStopList(stopTreeOrList)
        ? BASE_LINE_WIDTH + 1
        : xCoordForStop(stopTreeOrList, stopId);
      let coordinates = null;
      if (el) {
        const { offsetTop, offsetHeight } = el;
        const y = offsetTop + offsetHeight / 2;
        coordinates = [x, y];
      }

      dispatchStopCoords({
        type: "set",
        stop: stopId,
        coords: coordinates
      });
    });
  }, [dispatchStopCoords, stopRefsMap, stopTreeOrList]);

  useEffect(() => {
    updateAllStops();
    window.addEventListener("resize", updateAllStops);
    return () => window.removeEventListener("resize", updateAllStops);
  }, [updateAllStops]);

  return [stopRefsMap.current, updateAllStops];
}
