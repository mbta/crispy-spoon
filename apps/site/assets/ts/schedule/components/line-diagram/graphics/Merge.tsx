import { first, last } from "lodash";
import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { LineDiagramStop } from "../../__schedule";
import {
  getTreeDirection,
  isBranchTerminusStop,
  isMergeStop,
  lineDiagramIndexes,
  isStopOnMainLine
} from "../line-diagram-helpers";
import {
  CoordState,
  StopCoord,
  BASE_LINE_WIDTH,
  MERGE_RADIUS
} from "./graphic-helpers";
import Line from "./Line";

interface MergeGraphicsProps {
  lineDiagram: LineDiagramStop[];
}
const Merges = ({ lineDiagram }: MergeGraphicsProps): ReactElement | null => {
  const branchingOutward = getTreeDirection(lineDiagram) === "outward";
  const mergeIndices = lineDiagramIndexes(lineDiagram, isMergeStop);
  // walk through the diagram to find where all the merges go.
  const stopGaps = mergeIndices.map(i => {
    let from: LineDiagramStop;
    let to: LineDiagramStop;
    if (branchingOutward) {
      from = lineDiagram[i];
      [to] = lineDiagram.slice(i).filter(isStopOnMainLine);
    } else {
      to = lineDiagram[i];
      from = last(lineDiagram.slice(0, i - 1).filter(isStopOnMainLine))!;
    }

    return [from, to];
  });

  const lineDiagramCoords: CoordState = useSelector(
    (state: CoordState) => state
  );
  if (!lineDiagramCoords) return null;

  const mergeBends = stopGaps.map(([from, to]) => {
    // if outward, start with from and construct array of tos
    // if inward, start with to and construct array of froms
    const mergeStop = branchingOutward ? from : to;
    const mergeStopIndex = lineDiagram.indexOf(mergeStop);
    const mergeStopCoords = lineDiagramCoords[mergeStop.route_stop.id];
    if (!mergeStopCoords) return null;

    // usually there's just one terminus to draw the merge to... except for the GL at Kenmore.
    const mergeEnds = mergeStop.stop_data.slice(1).map(({ branch }) => {
      // along the branch might be subsequent stops or prior stops
      const downBranch = (branchingOutward
        ? lineDiagram.slice(mergeStopIndex)
        : lineDiagram.slice(0, mergeStopIndex - 1)
      ).filter(stop => stop.route_stop.branch === branch);
      const nextStopOnBranch = branchingOutward
        ? first(downBranch)
        : last(downBranch);
      const terminusOnBranch = branchingOutward
        ? first(downBranch.filter(isBranchTerminusStop))
        : last(downBranch.filter(isBranchTerminusStop));
      return [nextStopOnBranch, terminusOnBranch];
    });

    return mergeEnds.map(([nextStop, terminus]) => {
      if (!nextStop && !terminus) return null;
      // try the next stop on the branch first. if the branch is collapsed the
      // coordinates will be null as the next stop is hidden, so use terminus
      const nextStopVisible = !!lineDiagramCoords[nextStop!.route_stop.id];
      const nextStopCoords = nextStopVisible
        ? lineDiagramCoords[nextStop!.route_stop.id]
        : lineDiagramCoords[terminus!.route_stop.id];

      const [x, y] = nextStopCoords as StopCoord;
      const [mX, mY] = mergeStopCoords;
      const dx = (Math.abs(mX - x) - MERGE_RADIUS) * -1;
      const dy =
        (Math.abs(mY - y) - MERGE_RADIUS - (branchingOutward ? 60 : 20)) *
        (branchingOutward ? -1 : 1);
      const arc = branchingOutward
        ? `-${MERGE_RADIUS},-${MERGE_RADIUS} 0 0 0 -${MERGE_RADIUS},-${MERGE_RADIUS}`
        : `-${MERGE_RADIUS},${MERGE_RADIUS} 0 0 1 -${MERGE_RADIUS},${MERGE_RADIUS}`;
      const d = `M${x},${y} v${dy} a${arc} h${dx}`;
      return (
        <path
          key={`${mergeStop.route_stop.id}-${terminus!.route_stop.id}-merge`}
          strokeWidth={`${BASE_LINE_WIDTH}px`}
          d={d}
        />
      );
    });
  });

  if (!stopGaps.length && !mergeBends.length) return null;

  return (
    <g className="line-diagram-svg__merge">
      {stopGaps.map(([from, to]) => (
        <Line
          key={`${from.route_stop.id}-${to.route_stop.id}-line`}
          from={from}
          to={to}
        />
      ))}
      {mergeBends}
    </g>
  );
};

export default Merges;
