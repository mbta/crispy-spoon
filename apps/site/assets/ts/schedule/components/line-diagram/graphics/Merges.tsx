import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import {
  isBranchingNode,
  isMergeNode,
  longestPath,
  nextStopIds,
  previousStopIds
} from "../../../../helpers/stop-tree";
import { hasAnActiveDiversion } from "../../../../models/alert";
import { Alert } from "../../../../__v3api";
import { StopId, StopTree } from "../../__schedule";
import {
  BASE_LINE_WIDTH,
  CoordState,
  MERGE_RADIUS,
  StopCoord
} from "./graphic-helpers";
import { Line } from "./Line";

interface Props {
  stopTree: StopTree;
  alerts: Alert[];
}

// [fromStopId, toStopId]
type StopGap = [StopId, StopId];

const notOnPrimaryPath = (primaryPath: StopId[]) => (stopId: StopId): boolean =>
  !primaryPath.includes(stopId);

const mergeStopGaps = (
  stopTree: StopTree,
  primaryPath: StopId[]
): StopGap[] => {
  const mergeStops: StopId[] = primaryPath.filter(id =>
    isMergeNode(stopTree, id)
  );
  return mergeStops.flatMap(toStopId => {
    const fromStopIds: StopId[] = previousStopIds(stopTree, toStopId).filter(
      notOnPrimaryPath(primaryPath)
    );
    return fromStopIds.map(fromStopId => [fromStopId, toStopId] as StopGap);
  });
};

const branchingStopGaps = (
  stopTree: StopTree,
  primaryPath: StopId[]
): StopGap[] => {
  const branchingStops: StopId[] = primaryPath.filter(id =>
    isBranchingNode(stopTree, id)
  );
  return branchingStops.flatMap(fromStopId => {
    const toStopIds: StopId[] = nextStopIds(stopTree, fromStopId).filter(
      notOnPrimaryPath(primaryPath)
    );
    return toStopIds.map(toStopId => [fromStopId, toStopId] as StopGap);
  });
};

const stopsToStart = (stopTree: StopTree, stopId: StopId): StopId[] => [
  ...previousStopIds(stopTree, stopId).flatMap(previousId =>
    stopsToStart(stopTree, previousId)
  ),
  stopId
];
const mergeBends = (
  stopTree: StopTree,
  coords: CoordState,
  mergeGaps: StopGap[],
  alerts: Alert[]
): (JSX.Element | null)[] =>
  mergeGaps.map(([lastStopOnBranch, trunkStop]) => {
    const trunkStopCoords: StopCoord | null = coords[trunkStop];
    if (!trunkStopCoords) return null;

    const branchStopIds = stopsToStart(stopTree, lastStopOnBranch);
    const startingStopId: StopId = branchStopIds[0];

    // Try the last stop on the branch first. If the branch is collapsed, the
    // coordinates will be null as the last stop is hidden, so use terminus.
    const branchStopCoords: StopCoord | null =
      coords[lastStopOnBranch] || coords[startingStopId];
    if (!branchStopCoords) return null;

    const [trunkX, trunkY] = trunkStopCoords;
    const [branchX, branchY] = branchStopCoords;
    const dx = (Math.abs(trunkX - branchX) - MERGE_RADIUS) * -1;
    const dy = (Math.abs(trunkY - branchY) - MERGE_RADIUS - 20) * 1;
    const arc = `-${MERGE_RADIUS},${MERGE_RADIUS} 0 0 1 -${MERGE_RADIUS},${MERGE_RADIUS}`;
    const strokeProp = hasAnActiveDiversion(lastStopOnBranch, alerts)
      ? { stroke: "url(#diagonalHatch)" }
      : {};
    const pathProps: React.SVGProps<SVGPathElement> = {
      key: `${startingStopId}-${trunkStop}-merge`,
      strokeWidth: `${BASE_LINE_WIDTH}px`,
      d: `M${branchX},${branchY} v${dy} a${arc} h${dx}`,
      ...strokeProp
    };
    return <path key={pathProps.key} {...pathProps} />;
  });

const stopsToTerminus = (stopTree: StopTree, stopId: StopId): StopId[] => [
  stopId,
  ...nextStopIds(stopTree, stopId).flatMap(nextId =>
    stopsToTerminus(stopTree, nextId)
  )
];
const branchingBends = (
  stopTree: StopTree,
  coords: CoordState,
  branchingGaps: StopGap[],
  alerts: Alert[]
): (JSX.Element | null)[] =>
  branchingGaps.map(([trunkStop, firstStopOnBranch]) => {
    const trunkStopCoords: StopCoord | null = coords[trunkStop];
    if (!trunkStopCoords) return null;

    const branchStopIds = stopsToTerminus(stopTree, firstStopOnBranch);
    const terminusStopId: StopId = branchStopIds[branchStopIds.length - 1];

    // Try the next stop on the branch first. If the branch is collapsed, the
    // coordinates will be null as the next stop is hidden, so use terminus.
    const branchStopCoords: StopCoord | null =
      coords[firstStopOnBranch] || coords[terminusStopId];
    if (!branchStopCoords) return null;

    const [trunkX, trunkY] = trunkStopCoords;
    const [branchX, branchY] = branchStopCoords;
    const dx = (Math.abs(trunkX - branchX) - MERGE_RADIUS) * -1;
    const dy = (Math.abs(trunkY - branchY) - MERGE_RADIUS - 60) * -1;
    const arc = `-${MERGE_RADIUS},-${MERGE_RADIUS} 0 0 0 -${MERGE_RADIUS},-${MERGE_RADIUS}`;
    const strokeProp = hasAnActiveDiversion(firstStopOnBranch, alerts)
      ? { stroke: "url(#diagonalHatch)" }
      : {};
    const pathProps: React.SVGProps<SVGPathElement> = {
      key: `${trunkStop}-${terminusStopId}-branch`,
      strokeWidth: `${BASE_LINE_WIDTH}px`,
      d: `M${branchX},${branchY} v${dy} a${arc} h${dx}`,
      ...strokeProp
    };

    return <path key={pathProps.key} {...pathProps} />;
  });

const Merges = ({
  stopTree,
  alerts
}: Props): ReactElement<HTMLElement> | null => {
  const coords: CoordState = useSelector((state: CoordState) => state);
  if (!coords) return null;

  const primaryPath = longestPath(stopTree);
  const mergeGaps: StopGap[] = mergeStopGaps(stopTree, primaryPath);
  const branchingGaps: StopGap[] = branchingStopGaps(stopTree, primaryPath);
  const stopGaps: StopGap[] = mergeGaps.concat(branchingGaps);
  const bends = mergeBends(stopTree, coords, mergeGaps, alerts).concat(
    branchingBends(stopTree, coords, branchingGaps, alerts)
  );

  if (!stopGaps.length && !bends.length) return null;

  return (
    <g className="line-diagram-svg__merge">
      {stopGaps.map(
        ([fromId, toId]) =>
          fromId &&
          toId && (
            <Line
              key={`${fromId}-${toId}-line`}
              stopTree={stopTree}
              fromId={fromId}
              toId={toId}
              alerts={alerts}
            />
          )
      )}
      {bends}
    </g>
  );
};

export default Merges;
