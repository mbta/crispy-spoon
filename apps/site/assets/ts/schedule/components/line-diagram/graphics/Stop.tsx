import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { StopId } from "../../__schedule";
import { StopCoord, CoordState, CIRC_RADIUS } from "./graphic-helpers";

const Stop = ({
  stopId,
  shuttle
}: {
  stopId: StopId;
  shuttle?: boolean;
}): ReactElement<SVGCircleElement> | null => {
  const coords: StopCoord | null = useSelector(
    (state: CoordState) => state[stopId]
  );
  if (!coords) return null;
  const [x, y] = coords;
  return (
    <circle
      className="line-diagram-svg__stop"
      r={`${CIRC_RADIUS}px`}
      cx={`${x}px`}
      cy={`${y}px`}
      data-shuttle={!!shuttle}
    />
  );
};

export default Stop;
