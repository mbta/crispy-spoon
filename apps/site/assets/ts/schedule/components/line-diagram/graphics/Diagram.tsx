import { max } from "lodash";
import React, { ReactElement } from "react";
import { LineDiagramStop } from "../../__schedule";
import {
  hasBranchLines,
  isBranchTerminusStop,
  areOnDifferentBranchLines,
  diagramWidth
} from "../line-diagram-helpers";
import Line from "./Line";
import Merges from "./Merge";
import Stop from "./Stop";
import { routeToModeName } from "../../../../helpers/css";
import { LiveDataByStop } from "../__line-diagram";
import VehicleIcons from "../VehicleIcons";

interface DiagramProps {
  lineDiagram: LineDiagramStop[];
  liveData: LiveDataByStop | null;
}

const Diagram = (props: DiagramProps): ReactElement<HTMLElement> | null => {
  const { lineDiagram, liveData } = props;
  const width = diagramWidth(
    max(lineDiagram.map(ld => ld.stop_data.length)) || 1
  );

  const vehicleIcons = liveData
    ? // eslint-disable-next-line @typescript-eslint/camelcase
      lineDiagram.map(({ route_stop: stop }, index) => {
        if (!liveData[stop.id]) return null;
        // Hide vehicles arriving to the origin from 'off the line'
        const vehicleData =
          index === 0
            ? liveData[stop.id].vehicles.filter(
                vehicle => vehicle.status === "stopped"
              )
            : liveData[stop.id].vehicles;

        return (
          <VehicleIcons
            key={`${stop.id}-vehicles`}
            stop={stop}
            vehicles={vehicleData}
          />
        );
      })
    : null;

  return (
    <>
      {vehicleIcons}
      <svg
        xmlns="http://www.w3.org/2000/svg"
        className={`line-diagram-svg ${routeToModeName(
          lineDiagram[0].route_stop.route!
        )}`}
        width={`${width + 4}px`}
        height="100%"
      >
        {/* If there are multiple branches, draw the lines and curves for those */
        hasBranchLines(lineDiagram) && <Merges lineDiagram={lineDiagram} />}

        {/* Draw lines between stops */
        lineDiagram.map((current, idx, self) => {
          if (self[idx + 1]) {
            if (
              isBranchTerminusStop(current) &&
              areOnDifferentBranchLines(current, self[idx + 1])
            )
              return null; // don't draw a line from a branch end to the main line

            return (
              <Line
                key={`${current.route_stop.id}-${self[idx + 1].route_stop.id}`}
                from={current}
                to={self[idx + 1]}
              />
            );
          }
          return null; // this is the end of the line, nothing to draw a line to
        })}

        {/* Draw circles for each stop */
        lineDiagram.map(stop => (
          <Stop key={stop.route_stop.id} stop={stop} />
        ))}
      </svg>
    </>
  );
};

export default Diagram;
