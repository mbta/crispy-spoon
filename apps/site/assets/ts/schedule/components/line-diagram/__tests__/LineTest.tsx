import { mount, ReactWrapper } from "enzyme";
import React from "react";
import * as redux from "react-redux";
import { aroundNow } from "../../../../models/__tests__/alert-test";
import { Alert, InformedEntitySet } from "../../../../__v3api";
import { RouteStop, StopTree } from "../../__schedule";
import { BASE_LINE_WIDTH } from "../graphics/graphic-helpers";
import { Line } from "../graphics/Line";
import { createStopTreeCoordStore } from "../graphics/useTreeStopPositions";

const routeStopA: RouteStop = { id: "a" } as RouteStop;
const routeStopB: RouteStop = { id: "b" } as RouteStop;
const routeStopC: RouteStop = { id: "c" } as RouteStop;
const routeStopD: RouteStop = { id: "d" } as RouteStop;
const routeStopE: RouteStop = { id: "e" } as RouteStop;

const stopTree: StopTree = {
  byId: {
    a: { id: "a", value: routeStopA },
    b: { id: "b", value: routeStopB },
    c: { id: "c", value: routeStopC },
    d: { id: "d", value: routeStopD },
    e: { id: "e", value: routeStopE }
  },
  edges: {
    a: { next: ["b"], previous: [] },
    b: { next: ["c", "d"], previous: ["a"] },
    c: { next: ["e"], previous: ["b"] },
    d: { next: [], previous: ["b"] },
    e: { next: [], previous: ["c"] }
  },
  startingNodes: ["a"]
};
const store = createStopTreeCoordStore(stopTree);

const [testX, testY] = [7, 17];
// mock the redux state
const mockState = {
  ["a"]: [testX, testY],
  ["b"]: [testX, testY + 7],
  ["c"]: [testX, testY + 14],
  ["d"]: [testX, testY + 14]
};

const alertA: Alert = {
  id: "MOCK-ALERT-A",
  severity: 7,
  priority: "high",
  lifecycle: "new",
  effect: "stop_closure",
  informed_entity: { stop: ["a"] } as InformedEntitySet,
  active_period: aroundNow()
} as Alert;
const alertB: Alert = {
  id: "MOCK-ALERT-B",
  severity: 7,
  priority: "high",
  lifecycle: "new",
  effect: "stop_closure",
  informed_entity: { stop: ["b"] } as InformedEntitySet,
  active_period: aroundNow()
} as Alert;

jest
  .spyOn(redux, "useSelector")
  .mockImplementation(selector => selector(mockState));

describe("Line component", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <svg>
          <Line stopTree={stopTree} fromId={"a"} toId={"b"} alerts={[]} />
        </svg>
      </redux.Provider>
    );
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(wrapper.exists("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("standard line has expected props", () => {
    const { x1, y1, x2, y2, strokeWidth } = wrapper
      .find("line.line-diagram-svg__line")
      .first()
      .props();
    expect(strokeWidth).toEqual(`${BASE_LINE_WIDTH}px`);
    expect(x1).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(x2).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(y1).toEqual(`${testY}px`);
    expect(y2).toEqual(`${testY + 7}px`);
  });
});

describe("Line component between stops with branches", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <Line stopTree={stopTree} fromId={"b"} toId={"d"} alerts={[]} />
      </redux.Provider>
    );
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(wrapper.exists("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("line on branch has expected props", () => {
    const { x1, y1, x2, y2, strokeWidth } = wrapper
      .find("line.line-diagram-svg__line")
      .last()
      .props();
    expect(strokeWidth).toEqual(`${BASE_LINE_WIDTH}px`);
    expect(x1).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(x2).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(y1).toEqual(`${testY + 7}px`);
    expect(y2).toEqual(`${testY + 14}px`);
  });
});

describe("Line component between stops with disruptions", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <Line
          stopTree={stopTree}
          fromId={"a"}
          toId={"b"}
          alerts={[alertA, alertB]}
        />
      </redux.Provider>
    );
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(wrapper.exists("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("has expected props, including stroke pattern", () => {
    const { x1, y1, x2, y2, strokeWidth, stroke } = wrapper
      .find("line.line-diagram-svg__line")
      .last()
      .props();
    expect(strokeWidth).toEqual(`${BASE_LINE_WIDTH}px`);
    expect(x1).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(x2).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(y1).toEqual(`${testY}px`);
    expect(y2).toEqual(`${testY + 7}px`);
    expect(stroke).toContain("url(#diagonalHatch)");
  });
});
