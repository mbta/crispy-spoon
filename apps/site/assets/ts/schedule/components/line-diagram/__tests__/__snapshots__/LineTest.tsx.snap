// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line component between stops with branches renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <Line stopTree={{...}} fromId=\\"b\\" toId=\\"d\\" alerts={{...}}>
    <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"24px\\" y2=\\"31px\\" />
  </Line>
</Provider>"
`;

exports[`Line component between stops with disruptions renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <Line stopTree={{...}} fromId=\\"a\\" toId=\\"b\\" alerts={{...}}>
    <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"17px\\" y2=\\"24px\\" stroke=\\"url(#diagonalHatch)\\" />
  </Line>
</Provider>"
`;

exports[`Line component renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <svg>
    <Line stopTree={{...}} fromId=\\"a\\" toId=\\"b\\" alerts={{...}}>
      <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"17px\\" y2=\\"24px\\" />
    </Line>
  </svg>
</Provider>"
`;
